pkg/graveler/settings/manager.go

```diff
@@ -0,0 +1,179 @@
+	}
+	return m
+}
+
+// Save persists the given proto.Message under the given repository and key.
+func (m *Manager) Save(ctx context.Context, repositoryID graveler.RepositoryID, key string, message proto.Message) error {
+	repo, err := m.refManager.GetRepository(ctx, repositoryID)
+	if err != nil {
+		return err
+	}
+	messageBytes, err := proto.Marshal(message)
+	if err != nil {
+		return err
+	}
+	err = m.blockAdapter.Put(ctx, block.ObjectPointer{
+		StorageNamespace: string(repo.StorageNamespace),
+		Identifier:       fmt.Sprintf(settingsSuffixTemplate, m.committedBlockStoragePrefix, key),
+		IdentifierType:   block.IdentifierTypeRelative,
+	}, int64(len(messageBytes)), bytes.NewReader(messageBytes), block.PutOpts{})
+	if err != nil {
+		return err
+	}
+	logger := logging.FromContext(ctx)
+	if logger.IsTracing() {
+		logger.
+			WithField("repo", repositoryID).
+			WithField("key", key).
+			WithField("setting", protojson.Format(message)).
+			Trace("saving repository-level setting")
+	}
+	return nil
+}
+
+// Get fetches the setting under the given repository and key, and returns the result.
+// The result is eventually consistent: it is not guaranteed to be the most up-to-date setting. The cache expiry period is 1 second.
+func (m *Manager) Get(ctx context.Context, repositoryID graveler.RepositoryID, key string, emptyMessage proto.Message) (proto.Message, error) {
+	k := cacheKey{
+		RepositoryID: repositoryID,
+		Key:          key,
+	}
+	message, err := m.cache.GetOrSet(k, func() (v interface{}, err error) {
+		messageBytes, err := m.getFromStore(ctx, repositoryID, key)
+		if err != nil {
+			return nil, err
+		}
+		message := proto.Clone(emptyMessage)
+		err = proto.Unmarshal(messageBytes, message)
+		if err != nil {
+			return nil, err
+		}
+		return message, nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	logFetchedSetting(logging.FromContext(ctx), repositoryID, key, message.(proto.Message))
+	return message.(proto.Message), nil
+}
+
+func (m *Manager) getFromStore(ctx context.Context, repositoryID graveler.RepositoryID, key string) ([]byte, error) {
+	repo, err := m.refManager.GetRepository(ctx, repositoryID)
+	if err != nil {
+		return nil, err
+	}
+	objectPointer := block.ObjectPointer{
+		StorageNamespace: string(repo.StorageNamespace),
+		Identifier:       fmt.Sprintf(settingsSuffixTemplate, m.committedBlockStoragePrefix, key),
+		IdentifierType:   block.IdentifierTypeRelative,
+	}
+	reader, err := m.blockAdapter.Get(ctx, objectPointer, -1)
+	if errors.Is(err, adapter.ErrDataNotFound) {
+		return nil, graveler.ErrNotFound
+	}
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		_ = reader.Close()
+	}()
+	return io.ReadAll(reader)
+}
+
+// UpdateWithLock atomically gets a setting, performs the update function, and persists the setting to the store.
+func (m *Manager) UpdateWithLock(ctx context.Context, repositoryID graveler.RepositoryID, key string, emptyMessage proto.Message, update func(message proto.Message)) error {
```

Comment:
```suggestion
func (m *Manager) UpdateWithLock(ctx context.Context, repositoryID graveler.RepositoryID, key string, messageTemplate proto.Message, update func(message proto.Message)) error {
```
might be a bit better, to explain how the default message is used.  Or document the parameter (and then the name matters less).  The thing is, the way I see it, the most important thing is _not_ that the message is empty, it's the _type_ of the message.